### DISCUSSION TITLE
Remove default syntax highlighting

### DISCUSSION THREAD

**Discussion by mikelu92** [emacs-tree-sitter/elisp-tree-sitter#161] [Q&A] [answered]
Hi,

I'm trying to figure out how to remove highlighting for variables. For context, In the go language, I just want to have the variable declarations highlighted. Anytime a variable is used though I want it to not be highlighted. In [the current highlighting code](https://github.com/tree-sitter/tree-sitter-go/blob/master/queries/highlights.scm) it highlights all variables (identifiers). Is there a way to remove the highlighting on identifiers?

I've figured out how to isolate just declarations as follows:

```
(tree-sitter-hl-add-patterns 'go                                         
  [(short_var_declaration left: (expression_list (identifier) @variable))
   (var_declaration (var_spec name: (identifier) @variable))])
```

but I don't know what capture name (if any) to use for just the `(identifier)` highlighting. Any tips?

**Comment by ubolonton** [comment, marked as answer]
We don't have a mechanism to selectively remove a specific highlighting pattern yet.

There are 2 things you can do at the moment:

1. Copy the query pattern file returned by `(tree-sitter-langs--hl-query-path 'go)`. Modify it to suit your need. Then advise the function `tree-sitter-langs--hl-query-path` to return that file instead. For example:

```emacs-lisp
(define-advice tree-sitter-langs--hl-query-path (:before-until (lang-symbol) override-default-patterns)
  (pcase lang-symbol
    ('go "/path/to/your/modified/highlights.scm")))
```

Not highlighting all identifiers as `@variable` is actually consistent with `tree-sitter-langs`'s styles. The Go patterns should be modified to do so. A pull request is welcome.

2. Add new patterns that override the default patterns. For example:

```emacs-lisp
;; tree-sitter-hl.el doesn't have this face, but it probably should.
(defface tree-sitter-hl-face:nil
  '((default :inherit default))
  ""
  :group 'tree-sitter-hl-faces)

(tree-sitter-hl-add-patterns 'go
  [(short_var_declaration left: (expression_list (identifier) @variable))
   (var_declaration (var_spec name: (identifier) @variable))
   ;; This is prioritized over the default from highlights.scm.
   (identifier) @nil])
```

**Reply by mikelu92** [reply]
hey @ubolonton I think some time after you posted this something was update that your `define-advice` snippet no longer works. Whever I open up a go file I get the following error:

```
tree-sitter-after-on-hook: (wrong-number-of-arguments ((t) (lang-symbol) (if (eq lang-symbol 'go) (let nil "/my/modified/highlights/go-highlights.scm"))) 2)
```

Could you advise on why this is happening now? Sorry I'm not that versed in elisp

**Reply by mikelu92** [reply]
nvm figured it out, just needed to change it to the following:

```
(define-advice tree-sitter-langs--hl-query-path (:before-until (lang-symbol &optional mode) override-default-patterns)
  (pcase lang-symbol
    ('go "/path/to/your/modified/highlights.scm")))
```
because that function was changed in the [following commit](https://github.com/emacs-tree-sitter/tree-sitter-langs/commit/4d23b6893a4149cf2fcd3cf38c175fb6b1044318)

**Comment by mikelu92** [comment]
@ubolonton thanks for the response! Is there something that documents `tree-sitter-langs` styles? In case I want to make a pr for go to match it completely

**Comment by ubolonton** [comment]
There is no documentation on that yet. The rule of thumb is following the docstrings of the faces defined in [tree-sitter-hl.el](https://github.com/ubolonton/emacs-tree-sitter/blob/master/lisp/tree-sitter-hl.el).

Some notable conventions are:
- Definitions and uses should be differentiated. Examples:
  + `@function` vs. `@function.call`.
  + `@method` vs. `@method.call`.
  + `@type.parameter` vs. `@type.argument`.
  + `@variable` vs. none: variable uses should not be highlighted.
- Special faces should have high priority (placed earlier in the pattern list): `@function.macro`, `@type.builtin`, `@variable.special`.
- "Contextual" patterns should have low priority (placed at the end). Example: string interpolation.

Note that the package `tree-sitter-langs` provides its own [highlighting patterns](https://github.com/ubolonton/tree-sitter-langs/tree/master/queries). [This page](https://ubolonton.github.io/emacs-tree-sitter/syntax-highlighting/) lists languages whose highlighting patterns have been substantially modified compared to upstreams'. You can use them as examples.
